rls/src/main/java/io/grpc/rls/LinkedHashLruCache.java

```diff
@@ -200,9 +201,10 @@ private V invalidate(K key, EvictionType cause) {
   }
 
   @Override
-  public final void invalidateAll(Iterable<K> keys) {
-    checkNotNull(keys, "keys");
+  public final void invalidateAll() {
     synchronized (lock) {
+      // make a copy to avoid ConcurrentModificationException
+      Iterable<K> keys = ImmutableSet.copyOf(delegate.keySet());
```

Comment:
I tried using a set iterator
```java
  public final void invalidateAll() {
    synchronized (lock) {
      Iterator<K> keyIterator = delegate.keySet().iterator();
      while (keyIterator.hasNext()) {
        K key = keyIterator.next();
        SizedValue existing = delegate.get(key);
        if (existing != null) {
          evictionListener.onEviction(key, existing, EvictionType.EXPLICIT);
        }
        keyIterator.remove();
      }
    }
  }
```

but unfortunately it still throws 

```
io.grpc.rls.CachingRlsLbClientTest > get_childPolicyWrapper_reusedForSameTarget FAILED
    java.util.ConcurrentModificationException
        at java.base/java.util.LinkedHashMap$LinkedHashIterator.remove(LinkedHashMap.java:732)
        at io.grpc.rls.LinkedHashLruCache.invalidateAll(LinkedHashLruCache.java:214)
        at io.grpc.rls.LinkedHashLruCache.close(LinkedHashLruCache.java:298)
        at io.grpc.rls.CachingRlsLbClient.close(CachingRlsLbClient.java:270)
        at io.grpc.rls.CachingRlsLbClientTest.tearDown(CachingRlsLbClientTest.java:175)

```

Reply:
And I tried `delegate.clear()` after loop
```java
  public final void invalidateAll() {
    synchronized (lock) {
      for (K key : delegate.keySet()) {
        SizedValue existing = delegate.get(key);
        if (existing != null) {
          evictionListener.onEviction(key, existing, EvictionType.EXPLICIT);
        }
      }
      delegate.clear();
    }
  }
```

It still fails
```
io.grpc.rls.CachingRlsLbClientTest > get_updatesLbState FAILED
    java.util.ConcurrentModificationException
        at java.base/java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:719)
        at java.base/java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:741)
        at io.grpc.rls.LinkedHashLruCache.invalidateAll(LinkedHashLruCache.java:207)
        at io.grpc.rls.LinkedHashLruCache.close(LinkedHashLruCache.java:296)
```

because `LinkedHashMap.get()` is modifying the map.

Reply:
Now I tried with `delegate.entrySet()`, then both worked.

Reply:
Neat! Thank you so much for the extra effort!